### PR TITLE
Loki: Send current time range when fetching labels and values

### LINKED ISSUE #26213: When I'm researching application logs in Loki, for some apps the auto-completion works and for others it doesn't
I'm not sure if this is a Grafana or a Loki bug, feel free to tell me to move it somewhere else if this is not the right place.

**What happened**:
If you start typing {container_name:"xxxx in the query bar, the auto-completion sometimes does not work. 
But if you type it all, and launch the query, results are returned so it suggests the autocomplete should have found the label. 

**What you expected to happen**:
I expect the auto-completion to work for every existing label values, at least for the selected time range.

**Anything else we need to know?**:
I believe it could be time related. The application that does not show up in the auto-completion has not reported anything in the last day. Even if I increase the time range in the top corner right before writing the query, it still not autocomplete.

**How to reproduce it (as minimally and precisely as possible)**:
Not quite sure, have a Grafana and a Loki running collecting logs of containers, and maybe see if it autocomplete works for  application that does not log frequently. 

**Environment**:
All the application lives in containers in Kubernetes
- Grafana version: Grafana v7.0.3 (00ee734baf)
- Data source type & version: Loki 1.5.0

Anyone else have this issue ?  
Thanks !

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
See https://github.com/grafana/grafana/issues/26213

This PR fixes the issue that current selected time range is not being passed to queries when fetching labels and autocomplete. I have decided to pass the time range to react editors (similarly as it is done in Angular editors), instead of absolute range in Explore and nothing in panels, so we keep this unified. This way, we can use the range in other editors as well to improve autocomplete. 

More notes in **Files changed** part of this PR. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/26213

### CODE REVIEW

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 1]
```diff
@@ -5,29 +5,15 @@ import _ from 'lodash';
...
 import { LokiQuery, LokiOptions } from '../types';
 import { LokiQueryField } from './LokiQueryField';
 import LokiExploreExtraField from './LokiExploreExtraField';
 
 type Props = ExploreQueryFieldProps<LokiDatasource, LokiQuery, LokiOptions>;
 
 export function LokiExploreQueryEditor(props: Props) {
-  const { query, data, datasource, history, onChange, onRunQuery } = props;
-
-  let absolute: AbsoluteTimeRange;
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx:17]: I have created a unified method for this and `calcAbsoluteRange` and moved it to `public/app/plugins/datasource/loki/query_utils.ts`

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx [hunk 2]
```diff
@@ -75,7 +61,7 @@ export class LokiQueryEditor extends PureComponent<Props, State> {
           onRunQuery={this.onRunQuery}
           history={[]}
           data={data}
-          absoluteRange={this.calcAbsoluteRange(data)}
+          absoluteRange={createAbsoluteRange(data)}
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:64]: I will have to do a check with platform team regarding if I can pass timerange from QueryRow to Editor.

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 3]
```diff
@@ -444,8 +444,8 @@ export default class LokiLanguageProvider extends LanguageProvider {
     return nanos ? Math.floor(nanos / NS_IN_MS / 1000 / 60 / 5) : 0;
   }
 
-  async getLabelValues(key: string): Promise<string[]> {
-    return await this.fetchLabelValues(key, this.initialRange);
+  async getLabelValues(key: string, absoluteRange = this.initialRange): Promise<string[]> {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/language_provider.ts:447]: Fetch label values based on time range. If no range provided, use initialRange.

File: public/app/plugins/datasource/loki/components/useLokiLabels.ts [hunk 4]
```diff
@@ -39,7 +39,7 @@ export const useLokiLabels = (
   };
 
   const tryLabelsRefresh = async () => {
-    await languageProvider.refreshLogLabels(absoluteRange);
+    await languageProvider.refreshLogLabels(absoluteRange, true);
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/useLokiLabels.ts:42]: We are setting `forceRefresh = true` and therefore we are fetching labels even though we didn't wait for 30s ( LABEL_REFRESH_INTERVAL.  We always want to have labels and values that are up to date with time range.

File: packages/grafana-data/src/types/datasource.ts [hunk 5]
```diff
@@ -293,6 +293,7 @@ export interface QueryEditorProps<
 > {
   datasource: DSType;
   query: TQuery;
+  range: TimeRange;
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/datasource.ts:296]: To Angular Query Editors, we are passing time range. However, we didn't do this for React query editors. In Explore, we were passing absolute range. I tried to unify this and pass range to Explore Editors, React Editors and Angular Editors, to keep it consistent. We need time range in some editors (Loki) to fetch labels/values based on selected time range.

File: public/app/features/dashboard/panel_editor/QueryEditorRow.tsx [hunk 6]
```diff
@@ -178,6 +178,7 @@ export class QueryEditorRow extends PureComponent<Props, State> {
           onChange={onChange}
           onRunQuery={this.onRunQuery}
           data={data}
+          range={getTimeSrv().timeRange()}
         />
```
Comment (ivanahuckova) [inline public/app/features/dashboard/panel_editor/QueryEditorRow.tsx:182]: Same as in Angular panels.

File: public/app/features/explore/QueryRow.tsx [hunk 7]
```diff
@@ -138,6 +138,7 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
           onBlur={noopOnBlur}
           onChange={this.onChange}
           data={queryResponse}
+          range={range}
           absoluteRange={absoluteRange}
```
Comment (ivanahuckova) [inline public/app/features/explore/QueryRow.tsx:142]: In Explore, we were passing range to Angular editors, now we are passing it everywhere.

File: packages/grafana-data/src/types/datasource.ts [hunk 8]
```diff
@@ -300,6 +300,7 @@ export interface QueryEditorProps<
    * Contains query response filtered by refId of QueryResultBase and possible query error
    */
   data?: PanelData;
+  range?: TimeRange;
   exploreId?: any;
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/datasource.ts:304]: Range prop is always sent to Angular editors, but sometimes it is beneficial to have it in React Editors as well (e.g. fetching labels based on current time range).

File: packages/grafana-data/src/types/datasource.ts [hunk 9]
```diff
@@ -322,7 +323,6 @@ export interface ExploreQueryFieldProps<
 > extends QueryEditorProps<DSType, TQuery, TOptions> {
   history: any[];
   onBlur?: () => void;
-  absoluteRange?: AbsoluteTimeRange;
   exploreId?: any;
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/datasource.ts:326]: This was used for Loki Editor, but we have replaced it with range. I have checked other Explore React Editors and I couldn't find any other use of absoluteRange besides that.

File: public/app/plugins/datasource/loki/components/useLokiLabels.ts (hunk 4 above)
Comment (aocenas) [reply public/app/plugins/datasource/loki/components/useLokiLabels.ts:42]: Do we have to use `force = true` here? Ideally condition inside would check if the range is different and invalidate the refresh interval. If the range does not change than we should probably still adhere to that interval so we do not spam the server unnecessarily.

File: public/app/plugins/datasource/loki/components/LokiQueryField.tsx [hunk 10]
```diff
@@ -5,14 +5,16 @@ import LokiLanguageProvider from '../language_provider';
...
 type LokiQueryFieldProps = Omit<
   LokiQueryFieldFormProps,
-  'syntax' | 'syntaxLoaded' | 'onLoadOptions' | 'onLabelsRefresh' | 'logLabelOptions'
+  'syntax' | 'syntaxLoaded' | 'onLoadOptions' | 'onLabelsRefresh' | 'logLabelOptions' | 'absoluteRange'
 >;
 
 export const LokiQueryField: FunctionComponent<LokiQueryFieldProps> = props => {
-  const { datasource, absoluteRange, ...otherProps } = props;
+  const { datasource, range, ...otherProps } = props;
+  const absoluteTimeRange = { from: range!.from!.valueOf(), to: range!.to!.valueOf() };
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiQueryField.tsx:13]: `range!.from!` is this just a problem with the typings or is it really possible for it to be undefined?

**Comment by aocenas** [comment]
Added some comments, otherwise I think the approach makes sense.

Wasn't there also same issue with Prometheus?

File: public/app/plugins/datasource/loki/components/LokiQueryField.tsx (hunk 10 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/LokiQueryField.tsx:13]: It is just a typing issue. We are passing range to `LokiQuertField` everywhere it is used. Wasn't sure if I should use `non-null operator`, or give "OR solution". But this seemed a better solution, as "OR solution" would never been used.

**Comment by ivanahuckova** [comment]
> Wasn't there also same issue with Prometheus?

Yes, I have created an issue https://github.com/grafana/grafana/issues/26670 and I'll work on it as soon as this PR merges, to keep PRs smaller. 🙂

File: public/app/plugins/datasource/loki/components/LokiQueryField.tsx (hunk 10 above)
Comment (aocenas) [reply public/app/plugins/datasource/loki/components/LokiQueryField.tsx:13]: Lets just add a comment here to mention that range.from is actually never optional.

**Review by aocenas — APPROVED** [review]

**Pull request merged**